### DISCUSSION TITLE
Passing custom props to <InnerBlocks.Content/>

### DISCUSSION THREAD

**Discussion by ghost** [WordPress/gutenberg#37415] [Q&A] [answered]
Is it possible to pass custom props to InnerBlocks/InnerBlocks.Content ? The below does not work. I want to send some props from my parent component to my child component. 

```
<div {...useBlockProps.save()}>
    <InnerBlocks.Content myCustomProp={myCustomPropValue}/>
</div>
```

**Comment by fabiankaegy** [comment, marked as answer]
@Jonofat Instead of using the `<InnerBlocks.Content />` component you can use the `useInnerBlockProps.save()` hook. 

You can see an example of it here: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/query/save.js

The `useInnerBlocksProps` hook allows you to pass custom attributes to your components

**Reply by ghost** [reply]
Thanks very much for responding. I don't really understand the example though unfortunately. 

My save function looks like this now but I don't understand how to pass `columns` as props?

```
	save: ({ attributes }) => {
		const { columns } = attributes;
		const blockProps = useBlockProps.save();
		const innerBlocksProps = useInnerBlocksProps.save(blockProps);

		return (
			<div className="container">
				<div className="row">
					<div {...innerBlocksProps}></div>
				</div>
			</div>
		);
	},
```

**Reply by fabiankaegy** [reply]
Sorry I'll add some more context here: 

The `useInnerBlocksProps` and `useInnerBlocksProps.save()` hooks take an object as the first parameter. That object does not need to be the object generated by the `useBlockProps` hook but it can be. 

The object that you pass in via the first property to the function can contain props that you want to be added to the inner blocks container. 

In your example, you likely want something like this:

```js
save: ({ attributes }) => {
   const { columns } = attributes;
   const innerBlocksProps = useInnerBlocksProps.save({ dataColumns: columns });

return (
	<div className="container">
		<div className="row">
			<div {...innerBlocksProps} />
		</div>
	</div>
);
},
```

**Reply by ghost** [reply]
Ah, thanks for the explanation! 

I have tried this but am getting an error from this line according to console.

`const innerBlocksProps = useInnerBlocksProps.save({ dataColumns: columns });`

Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'save')

**Reply by fabiankaegy** [reply]
2 questions:

Are you setting `apiVersion` to 2 in the `registerBlockType` function or `block.json` file? 
Are you running the Gutenberg Plugin on your site? In WordPress core the `useInnerBlocksProps` function is still marked as experimental and needs to be imported like so:

```js
import { __experimentalUseInnerBlocksProps as useInnerBlocksProps } from '@wordpress/block-editor';
```

**Reply by ghost** [reply]
Yes, in my block.json file I have set the api version to 2. Not sure what you mean by running the Gutenberg plugin on the site? The Gutenberg Editor is just standard with the latest version of WordPress I have installed but not sure if that is the same thing you are referring to. 

After changing the import as per your code the error is gone! The last thing I think though is how to access the dataColumns object. In the child component when I console.log props I don't see it there?

**Reply by fabiankaegy** [reply]
Gutenberg itself is a Feature Plugin that you can install. This repository here on GitHub contains the latest code for this plugin. 
Features that get developed in the plugin first then get released in WordPress Core whenever new versions are released. At the moment the `useInnerBlocksProps` hook is only stable in the Gutenberg Plugin and will be released as a stable API in WordPress 5.9.


And I'm sorry I misunderstood your question slightly. I thought your goal was to add a data attribute to the wrapping element so that you could use it on the frontend etc.

If you want to pass down attributes to child blocks you can do that through Block Context. You can find the documentation for that there: https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-context.md

**Reply by ghost** [reply]
Thanks, I will take a look at this but what I ended up doing in the meantime which is probably overkill for what I am trying to do is to use useSelect and useDispatch with my own redux store. So, if I update the number of columns in the parent component I get that value in the child component from redux. Would have been a lot nicer to just pass the number of columns value down with props but perhaps the Context example will be a better alternative.
